libbeat/processors/xmldecode/xmldecode_test.go

```diff
@@ -0,0 +1,255 @@
+			description: "Simple xml decode with xml string to same field name when Target is null",
+			config: xmlDecodeConfig{
+				Fields: []string{"message"},
+			},
+			Input: common.MapStr{
+				"message": `<?xml version="1.0"?>
+				<catalog>
+					<book seq="1">
+						<author>William H. Gaddis</author>
+						<title>The Recognitions</title>
+						<review>One of the great seminal American novels of the 20th century.</review>
+					</book>
+				</catalog>`,
+			},
+			Output: common.MapStr{
+				"message": map[string]interface{}{
+					"catalog": map[string]interface{}{
+						"book": map[string]interface{}{
+							"author": "William H. Gaddis",
+							"review": "One of the great seminal American novels of the 20th century.",
+							"seq":    "1",
+							"title":  "The Recognitions",
+						},
+					},
+				},
+			},
+			error:        false,
+			errorMessage: "",
+		},
+		{
+			description: "Decoding with array input",
+			config: xmlDecodeConfig{
+				Fields: []string{"message"},
+			},
+			Input: common.MapStr{
+				"message": `<?xml version="1.0"?>
+				<catalog>
+					<book>
+						<author>William H. Gaddis</author>
+						<title>The Recognitions</title>
+						<review>One of the great seminal American novels of the 20th century.</review>
+					</book>
+					<book>
+						<author>Ralls, Kim</author>
+						<title>Midnight Rain</title>
+						<review>Some review.</review>
+					</book>
+				</catalog>`,
+			},
+			Output: common.MapStr{
+				"message": map[string]interface{}{
+					"catalog": map[string]interface{}{
+						"book": []interface{}{
+							map[string]interface{}{
+								"author": "William H. Gaddis",
+								"review": "One of the great seminal American novels of the 20th century.",
+								"title":  "The Recognitions",
+							},
+							map[string]interface{}{
+								"author": "Ralls, Kim",
+								"review": "Some review.",
+								"title":  "Midnight Rain",
+							},
+						},
+					},
+				},
+			},
+			error:        false,
+			errorMessage: "",
+		},
+		{
+			description: "Decoding with multiple xml objects",
+			config: xmlDecodeConfig{
+				Fields: []string{"message"},
+			},
+			Input: common.MapStr{
+				"message": `<?xml version="1.0"?>
+				<catalog>
+					<book>
+					<author>William H. Gaddis</author>
+					<title>The Recognitions</title>
+					<review>One of the great seminal American novels of the 20th century.</review>
+				</book>
+				<book>
+					<author>Ralls, Kim</author>
+					<title>Midnight Rain</title>
+					<review>Some review.</review>
+				</book>
+				<secondcategory>
+					<paper id="bk102">
+						<test2>Ralls, Kim</test2>
+						<description>A former architect battles corporate zombies, 
+						an evil sorceress, and her own childhood to become queen 
+						of the world.</description>
+					</paper>
+				</secondcategory>
+				</catalog>`,
+			},
+			Output: common.MapStr{
+				"message": map[string]interface{}{
+					"catalog": map[string]interface{}{
+						"book": []interface{}{
```

Comment:
Oh, one important thing I have just recalled. This case of arrays of objects won't be actually work as expected, this is not supported in Elasticsearch, see the warning in https://www.elastic.co/guide/en/elasticsearch/reference/7.10/array.html

You can read more about this in the JSON case on this comment: https://github.com/elastic/beats/pull/21178#issuecomment-705464067

The solution would probably be to generate multiple events, but this cannot be done with the current processors framework. It can be done at the processor level, for example the `httpjson` input has options to support these cases generating multiple events from single objects or arrays.

I mean, not much change here, but just take into account that a processor shouldn't be used to decode arrays of objects. Only thing would be to modify the test cases to include examples with single objects, that is also the main use case for `decode_json_fields`.

Reply:
Array of objects would still be possible to ingest and visualize in some manners, I don't think we should error on it or not allow it at all.

The processor will just decode what is sent to it, and indeed the splitting will happen on the input side.
An array of objects still has certain usecases, and they might also want to include it in the pipeline solely so that you can use ingest nodes to iterate over them and for example append them to a field.

The test cases already covers both single and multiple object usecases, would there be any reason not to @jsoriano ?

Reply:
Just to clarify, arrays of objects are perfectly fine and can even be queried independently if you're working with `nested` or `flattened` types -- I can very much imagine that as a use-case for this processor if you're wanting to model an event with nested information. So I don't think we should necessarily give any guidance that you shouldn't use this processor for arrays of data, it all just depends on what you're modeling.

- If you want to model an array as multiple events, use the input
- If you want to model `nested` or `flattened` fields, this processor would work great for that

Reply:
Agree, my comment was more for ourselves, to be clear that arrays of objects is a use case that usually don't work as expected, without a lot of care.
Mappings with `nested` or `flattened` fields effectively would work, but this is not something that we are frequently using in beats, and is something not so straightforward to use in general.

Reply:
I didn't know there have been some improvements in kibana during 7.x for nested fields https://github.com/elastic/kibana/issues/1084#issuecomment-585178079, perhaps we could use them more, yes.